include/deal.II/base/aligned_vector.h

```diff
@@ -259,6 +260,93 @@ class AlignedVector
+   *
+   * @note The use of shared memory between MPI processes requires
+   *   that the detected MPI installation supports the necessary operations.
+   *   This is the case for MPI 3.0 and higher.
+   *
+   * @note This function is not cheap. It needs to create sub-communicators
+   *   of the provided @p communicator object, which is generally an expensive
+   *   operation. Likewise, the generation of shared memory spaces is not
+   *   a cheap operation. As a consequence, this function primarily makes
+   *   sense when the goal is to share large read-only data tables among
+   *   processes; examples are data tables that are loaded at start-up
+   *   time and then used over the course of the run time of the program.
+   *   In such cases, the start-up cost of running this function can be
+   *   amortized over time, and the potential memory savings from not having to
+   *   store the table on each process may be substantial on machines with
+   *   large core counts on which many MPI processes run on the same machine.
+   *
+   * @note This function only makes sense if the data type `T` is
+   *   "self-contained", i.e., all if its information is stored in its
+   *   member variables, and if none of the member variables are pointers
+   *   to other parts of the memory. This is because if a type `T` does
+   *   have pointers to other parts of memory, then moving `T` into
+   *   a shared memory space does not result in the other processes having
+   *   access to data that the object points to with its member variable
+   *   pointers: These continue to live only on one process, and are
+   *   typically in memory areas not accessible to the other processes.
+   *   As a consequence, the usual use case for this function is to share
+   *   arrays of simple objects such as `double`s or `int`s.
+   *
+   * @note After calling this function, objects on different MPI processes
+   *   share a common state. That means that certain operations become
+   *   "collective", i.e., they must be called on all participating
+   *   processors at the same time. In particular, you can no longer call
+   *   resize(), reserve(), or clear() on one MPI process -- you have to do
+   *   so on all processes at the same time, because they have to communicate
+   *   for these operations. If you do not do so, you will likely get
+   *   a deadlock that may be difficult to debug. By extension, this rule of
+   *   only collectively resizing extends to this function itself: You can
+   *   not call it twice in a row because that implies that first all but the
+   *   `root_process` throw away their data, which is not a collective
+   *   operation. Generally, these restrictions on what can and can not be
+   *   done hint at the correctness of the comments above: You should treat
+   *   an AlignedVector on which the current function has been called as
+   *   `const`, on which no further operations can be performed until
+   *   the destructor is called.
```

Comment:
Maybe nothing. It is true that we could go into much more detail as for one *could* in principle do after calling `replicate_across_communicator()`, but my goal with this description is to not do that: I don't want to support any operations after the call, and got bitten several times in attempts at implementing any other scheme that would provide semantics to what these functions beyond declaring that one should consider the object as `const`. I also see no reason to support anything other: if you want another object, just create a new one.

Reply:
If this is the way users should use this class maybe we should disable the mentioned functions (except of `clear()`) by adding some asserts that are triggered only if `replicate_across_communicator()` was called in debug mode.

Reply:
I thought about that as well, but it would require storing information somewhere in a flag -- and making the object larger by another 4 bytes. That was already the most contentious point :-(

For completeness, I had also considered storing this information elsewhere. For example, we have the two pointers that point at the end of the *used* area and the end of the *allocated* area. I had thought about setting the latter to `nullptr` to indicate that we have a read-only object. That didn't seem like a particularly nice way to do that, but has the advantage that we don't waste memory. If that would be acceptable, then I'd be happy to do this.